elb_application_lb: fix issue with botocore parameter validation on Rules. Fixes #27332

mtinning opened the pull request:
##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->
Fix to issue in `lib\ansible\modules\cloud\amazon\elb_application_lb.py` that was causing `ParameterValidationException` in `boto`.

Removing invalid parameters from listener args before calling `boto`'s `connection.create_listener` or `connection.modify_listener`. Ensuring easy updating by checking against list of known valid values.

Added tests.

Fixes #27332 

<!---
If you are fixing an existing issue, please include "Fixes #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Bugfix Pull Request

##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
`elb_application_lb`

##### ANSIBLE VERSION
<!--- Paste verbatim output from “ansible --version” between quotes below -->
```
ansible 2.4.0 (issue-27332-elb-application-lb-param-validation 476893ca8b) last updated 2017/07/26 17:05:18 (GMT +000)
  config file = None
  configured module search path = [u'/root/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /ansible/.ansible-fork/lib/ansible
  executable location = /ansible/.ansible-fork/bin/ansible
  python version = 2.7.12 (default, Nov 19 2016, 06:48:10) [GCC 5.4.0 20160609]
```


##### ADDITIONAL INFORMATION
<!---
Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful.
  -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
See #27332

mtinning committed:
elb_application_lb: fix issue with boto parameter validation on Rules

abadger commented on lib/ansible/modules/cloud/amazon/elb_application_lb.py:
```diff
@@ -399,6 +399,12 @@ def _get_subnet_ids_from_subnet_list(subnet_list):
+def get_valid_listener_parameters(listener):
+    # See CreateListenerInput in https://github.com/boto/botocore/blob/f51271ac3ce4947ab3bbe46b82bdd79d8be0d5c2/botocore/data/elbv2/2015-12-01/service-2.json
+    valid_parameters = ['LoadBalancerArn', 'Protocol', 'Port', 'DefaultActions', 'SslPolicy', 'Certificates']
+    return {key: value for key, value in listener.items() if key in valid_parameters}
```
Just a driveby comment: for pyhton-2.6 compatibility we cannot use dict comprehensions.  However, this syntax is almost equivalent:
```
return dict((key, value) for key, value in listener.items() if key in valid_parameters)
```

wimnat commented:
This just negates user error.  The reason the error is thrown is because the user entered an InvalidParameter (Rules).  By filtering out all invalid parameters with this function you could be hiding relevant errors from users.  I would just change this to remove the `Rules` key before update is called.

mtinning commented:
@wimnat `Rules` is listed as a valid parameter in one of the examples (see [Line 177](https://github.com/mtinning/ansible/blob/dc3e98ff39cdb9361f68037acf08b302bc538b5b/lib/ansible/modules/cloud/amazon/elb_application_lb.py#L177))

In addition, in order to add rules to the ELB `'Rules'` is expected to be present in the parameters the user defines (see [Line 753](https://github.com/mtinning/ansible/blob/dc3e98ff39cdb9361f68037acf08b302bc538b5b/lib/ansible/modules/cloud/amazon/elb_application_lb.py#L753), amongst others).

I agree that `'Rules'` is an invalid parameter to pass to boto, however given the current design I disagree that this is user error as the module specification is being followed.

What do you mean by:

> I would just change this to remove the `Rules` key before update is called.

Which "update" are you referring to?

s-hertel commented:
Hi @mtinning. The issue with this is that listeners won't be updated with rules upon creation. You'd have to run the same playbook twice to get the desired result. The parameter is valid, just not for creation, so it needs to be temporarily removed. Then the created listener needs to be added to the listeners to be updated. I've made an alternative fix here: #27379

- edit
...and actually, since you've added that parameter validation to modify listeners, I don't think the listener will get the appropriate rules even after rerunning the playbook.

mtinning commented:
@s-hertel thanks for your comment.

The `Rules` parameter is valid for neither creation nor modification of the ELB - see the [parameters listed in the docs](http://boto3.readthedocs.io/en/latest/reference/services/elbv2.html#ElasticLoadBalancingv2.Client.modify_listener). As a result I don't think your #27379 fixes the issue completely.

> You'd have to run the same playbook twice to get the desired result.

I don't agree with this. There is an issue that after creating the listener the `ListenerArn` has not been set - I made a change locally to get this from the value returned from `create_listener`, which I will put up for PR once this has been approved.

s-hertel commented:
@mtinning Thanks for the reply.

Sorry, I could have chosen my wording more carefully. I did not mean boto3's modify_listener() function, but that modification to the listeners happens after verifying that the desired ones are present. Here are the docs responsible for adding the rules:
http://boto3.readthedocs.io/en/latest/reference/services/elbv2.html#ElasticLoadBalancingv2.Client.create_rule
Here's where the rules are used:
https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/cloud/amazon/elb_application_lb.py#L743-L778

Since it requires the listener's ARN it cannot be done during creation. It would be awesome if you could test my PR to see if it solves your issue correctly, because it may be wrong.

> I made a change locally to get this from the value returned from create_listener, which I will put up for PR once this has been approved.

That fix is also in my PR.

s-hertel commented:
@mtinning I checked out your change again with more of an open mind (I disregarded it when I was met with an ugly traceback, I'm afraid), and the way you're approaching this seems totally fine. Once you add the created listener to the listeners list for future modification I'll review this. Also, for your tests you could use the placebo fixtures. Not sure how helpful it is just to test that the functions are being called with appropriate arguments, but it would be great to have some tests (integration tests also an option).

s-hertel commented:
(another musing - you'll also need to update the current_listeners in addition to adding the created listener to listeners.)

wimnat commented:
I still don't think this is an appropriate fix.  If i add in some random parameter to my listener then i'll never be notified i've done something wrong.

E.g.

`
elb_application_lb:
  listeners:
    stawberries: "bla"
    ...
  ...
`

The above task would run without error.  I think it's important to tell the user that `stawberries` is an invalid parameter.

As I said, i think a better approach is just to pop the Rules before you call create or modify.

> I don't agree with this. There is an issue that after creating the listener the ListenerArn has not been set - I made a change locally to get this from the value returned from create_listener, which I will put up for PR once this has been approved.

@mtinning - Can you expand on why this is necessary?  Before creating rules i call out to get the ListenerArn - https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/cloud/amazon/elb_application_lb.py#L751

mtinning commented:
@wimnat no dispute from me on the merit of `pop`ing `Rules` instead of removing all invalid parameters.

> I don't agree with this. There is an issue that after creating the listener the ListenerArn has not been set - I made a change locally to get this from the value returned from create_listener, which I will put up for PR once this has been approved.

The `ListenerArn` is needed before rule creation, in `compare_rules` - https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/cloud/amazon/elb_application_lb.py#L675. This is the line that is causing the issue.

I'll make the requested change.

mtinning commented:
@s-hertel thanks for taking another look. I appreciate an initially failing build producing an "ugly traceback" makes for an unappealing PR, however if the barrier to entry was ensuring a fully working build before making a PR I wouldn't be in a position to contribute. I'll make the `ListenerArn` update.

mtinning commented:
I've now added the fix for the missing `ListenerArn` issue (the new listener has been added to `current_listeners`.

@s-hertel I spent some time looking at the `placebo` module and the `placeboify` fixture. They look excellent, and I agree they would make for much better tests. However I can't currently spare the time to set them up so I have stuck with the current approach. There are obvious drawbacks, as the mocks take a bit of effort to set up and the tests are currently quite strongly coupled to the internal structure of the module, but I think some tests is better than no tests. If you disagree I can remove them.

@wimnat I changed the implementation to just `pop` the `Rules` parameter.

Thanks for all your comments.

wimnat commented:
Thanks @mtinning. Just fix up ansible test sanity errors and it LGTM. shipit

mtinning commented:
@wimnat thanks - sanity tests now passing (unless I've missed something?)

wimnat commented:
LGTM

s-hertel commented on lib/ansible/modules/cloud/amazon/elb_application_lb.py:
```diff
@@ -719,14 +719,20 @@ def create_or_update_elb_listeners(connection, module, elb):
         try:
             listener_to_add['LoadBalancerArn'] = elb['LoadBalancerArn']
-            connection.create_listener(**listener_to_add)
+            # Rules is not a valid parameter for create_listener
```
Looks good! On line [744](https://github.com/ansible/ansible/pull/27333/files#diff-a1bcebad578f0e6bfd841049d327a587L744) can you please also fix that to be deepcopy(listeners) (so the invocation gets shown correctly)? 
And lines [691](https://github.com/ansible/ansible/pull/27333/files#diff-a1bcebad578f0e6bfd841049d327a587L691) and 692 need to be set equal for new_rule, not current_rule. After that is appears rules are being updated correctly for me. Thanks for working on this.

mtinning replied:
Good spot for lines 691 & 692 - I'll fix this.

Hope I'm not being dim, but what does "so the invocation gets shown correctly" refer to? The change you suggest for line 744 seems harmless but I don't understand the reason for it at the moment.

Thanks!

s-hertel replied:
When a task is run with extra verbosity it should display the arg spec the module used after the 'invocation' key (run playbook with -vvv). I could see that somewhere that actual dictionary was being modified, rather than a copy of it. 
I thought I kept running the wrong playbook or was doing something wrong syntactically because it was deleting my rules:
```
...
            "listeners": [
                {
                    "DefaultActions": [
                        {
                            "TargetGroupArn": "arn:aws:elasticloadbalancing:us-east-1:448830907657:targetgroup/sherteltg/bd83c5c780de26d7",
                            "Type": "forward"
                        }
                    ],
                    "ListenerArn": "arn:aws:elasticloadbalancing:us-east-1:448830907657:listener/app/shertel-application-lb/365334adfd1f3606/9118ad2fccc38cde",
                    "Port": 80,
                    "Protocol": "HTTP",
                    "Rules": []
                }
            ],
```

mtinning commented:
@s-hertel pushed up the requested changes so that I can knock this off before the weekend.

Thanks!

ryansb commented on test/units/modules/cloud/amazon/test_elb_application_lb.py:
```diff
@@ -0,0 +1,166 @@
+#
+# (c) 2017 Michael Tinning
```
Copyright guidelines have changed recently to be much shorter (and easier to follow) [here](http://docs.ansible.com/ansible/latest/dev_guide/developing_modules_documenting.html#copyright)

So you can delete the whole license snippet below this and add:

```
# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
```

ryansb commented on test/units/modules/cloud/amazon/test_elb_application_lb.py:
```diff
@@ -0,0 +1,166 @@
+from ansible.module_utils.ec2 import HAS_BOTO3
+
+if not HAS_BOTO3:
+    raise SkipTest("test_api_gateway.py requires the `boto3` and `botocore` modules")
```
s/test_api_gateway.py/test_elb_application_lb.py/

ryansb commented on test/units/modules/cloud/amazon/test_elb_application_lb.py:
```diff
@@ -0,0 +1,166 @@
+    connection,
+    listener,
+    elb,
+    compare_listeners,
```
You don't need to re-include some of these fixtures, as they are used in the `elb` fixture but never called in this test.

mtinning replied:
All of these fixtures are called in this test (believe it or not!)

mtinning commented:
@ryansb thanks for your comments - I've made the requested changes

mtinning commented:
@ryansb is there any chance we can try to get this in? Is there anything else you'd like to request? I'd like to be working off the main ansible devel branch as fixes are rolling in that I need, but this issue prevents that. Many thanks.

ryansb commented:
rebuild_merge
